[ACP-77] Update Continuous Fee Mechanism (#119)
avalanche-foundation/ACPs: ACPs/77-reinventing-subnets/README.md

        self.queue.add(vdr)
```

#### Fee Algorithm

[ACP-103](../103-dynamic-fees/README.md) proposes a dynamic fee mechanism for transactions on the P-Chain. This mechanism is repurposed with minor modifications for the active Subnet Validator continuous fee.

At the start of building/executing block $b$, the number of excess active Subnet Validators $x$ is updated:

$$x = \max(x + (V - T) \cdot \Delta t, 0)$$

Where:

- $V$ is the number of active Subnet Validators prior to the execution of block $b$
- $T$ is the target number of active Subnet Validators
- $\Delta t$ is the number of seconds between $b$'s block timestamp and $b$'s parent's block timestamp

The fee per Subnet Validator from $b$'s parent's block timestamp to $b$'s block timestamp is:

> $$M \cdot \exp\left(\frac{x}{K}\right)$$

Reviewer: Doesn't there need to be some notion of time here?

Whenever $x$ increases by $K$, the price per active Subnet Validator increases by a factor of `~2.7`. If the price per active Subnet Validator gets too expensive, some active Subnet Validators will exit the active validator set, decreasing $x$, dropping the price. The price per active Subnet Validator constantly adjusts to make sure that, on average, the P-Chain has no more than $T$ active Subnet Validators.

The parameters at activation are:

| Parameter | Value |
| - | - |
| $T$ | `10000` Subnet Validators |
| $C$ | `20000` Subnet Validators |
| $M$ | `2048` nAVAX/s (`0.1769472` AVAX/day) |
| $K$ | `60_480_000_000` |

These parameters air on the conservative side as a precautionary measure to ensure network stability during activation. `60_480_000_000` was chosen for $K$ to have the validator fee increase by a factor of `~2.7` after a full week of being at $C$ or `20000` Subnet Validators. At activation, this should strike a good balance between preventing squatting and spikiness of the continuous fee.

A future ACP can adjust the parameters to increase $T$, reduce $M$, and/or modify $K$.

#### Block Timestamp Validity Change

> The above algorithm assumes that the difference between $V$ and $T$ is constant in the $\Delta T$ between blocks. However, this isn't always true.

Reviewer: ```suggestion
The above algorithm assumes that the difference between $V$ and $T$ is constant in the $\Delta T$ between blocks. Therefore, it must be ensured that the block timestamp does not advance past the time that the first eviction can occur.
```

> Suppose the network currently has `11000` Subnet Validators. With $T$ being `10000`, $x$ will increase by $V-T$ or `1000` every second. If the next block's timestamp is `5` seconds after the current block's timestamp. $x$ will increase by `5000`. This is expected behavior _unless_ multiple Subnet Validators get removed within those `5` seconds. If `100` Subnet Validators are removed `1` second after the current block timestamp, $x$ should only increase by `900` every second after the first second elapses. $x$ should have increased by `1000` + `4 * 900` = `4500`, not `5000`.

Reviewer: ```suggestion
Suppose the network currently has `11000` Subnet Validators. With $T$ being `10000`, $x$ will increase by $V-T$ or `1000` every second. If the next block's timestamp is `5` seconds after the current block's timestamp. $x$ will increase by `5000`. This is expected behavior _unless_ Subnet Validators should have been removed within those `5` seconds. If `100` Subnet Validators would exhaust their balance `1` second after the current block timestamp, $x$ should only increase by `900` every second after the first second elapses. $x$ should have increased by `1000` + `4 * 900` = `4500`, not `5000`.
```

To ensure that the correct fee is always charged per unit of time, blocks are considered valid if their timestamps are no greater than the time at which the first Subnet Validator gets removed from a lack of funds. This upholds the invariant that the number of Subnet Validators remains constant between blocks.

> The block building protocol is modified to account for this change by first checking if the wall clock time removes any Subnet Validator due to a lack of funds. If the wall clock time does remove any Subnet Validators, the wall clock time is used to build the block. If it does, a binary search is conducted between the parent block's timestamp and the wall clock time to find the time at which the first Subnet Validator gets removed.

Author: ```suggestion
The block building protocol is modified to account for this change by first checking if the wall clock time removes any Subnet Validator due to a lack of funds. If the wall clock time does not remove any Subnet Validators, the wall clock time is used to build the block. If it does, a binary search is conducted between the parent block's timestamp and the wall clock time to find the time at which the first Subnet Validator gets removed.
```

> The block building protocol is modified to account for this change by first checking if the wall clock time removes any Subnet Validator due to a lack of funds. If the wall clock time does not remove any Subnet Validators, the wall clock time is used to build the block. If it does, a binary search is conducted between the parent block's timestamp and the wall clock time to find the time at which the first Subnet Validator gets removed.

Reviewer: ```suggestion
The block building protocol is modified to account for this change by first checking if the wall clock time removes any Subnet Validator due to a lack of funds. If the wall clock time does not remove any Subnet Validators, the wall clock time is used to build the block. If it does, a binary search may be performed between the parent block's timestamp and the wall clock time to find the time at which the first Subnet Validator gets removed.
```

  ```python
  # Approximates factor * e ** (numerator / denominator) using Taylor expansion
  def fake_exponential(factor: int, numerator: int, denominator: int) -> int:
    i = 1
    output = 0
    numerator_accum = factor * denominator
    while numerator_accum > 0:
        output += numerator_accum
        numerator_accum = (numerator_accum * numerator) // (denominator * i)
        i += 1
    return output // denominator
  ```

- $K$ is a constant to control the rate of change for the Subnet Validator price

After every second, $x$ will be updated:

$$x = \max(x + (V - T), 0)$$

Where:

- $V$ is the number of active Subnet Validators
- $T$ is the target number of active Subnet Validators

Whenever $x$ increases by $K$, the price per active Subnet Validator increases by a factor of `~2.7`. If the price per active Subnet Validator gets too expensive, some active Subnet Validators will exit the active validator set, decreasing $x$, dropping the price. The price per active Subnet Validator constantly adjusts to make sure that, on average, the P-Chain has no more than $T$ active Subnet Validators.

#### Block Timestamp Validity Change

> To easily track $x$ between blocks, the difference between $V$ and $T$ must be constant in the $\Delta t$ between blocks. $x$ can then simply be updated to $\max(x + \Delta t \cdot (V - T), 0)$ after each valid block. If $V-T$ is not constant between blocks, this optimization cannot be made.
> 
> Suppose the network currently has `11000` Subnet Validators. With $T$ being `10000`, $x$ will increase by $V-T$ or `1000` every second. If the next block's timestamp is `5` seconds after the current block's timestamp. $x$ will increase by `5000`. This is expected behavior _unless_ Subnet Validators should have been removed within those `5` seconds. If `100` Subnet Validators would exhaust their balance `1` second after the current block timestamp, $x$ should only increase by `900` every second after the first second elapses. $x$ should have increased by `1000` + `4 * 900` = `4500`, not `5000`.
> 
> To ensure that the correct fee is always charged per unit of time, blocks are considered valid if their timestamps are no greater than the time at which the first Subnet Validator gets removed from a lack of funds. This upholds the invariant that the number of Subnet Validators remains constant between blocks.
> 
> The block building protocol is modified to account for this change by first checking if the wall clock time removes any Subnet Validator due to a lack of funds. If the wall clock time does not remove any Subnet Validators, the wall clock time is used to build the block. If it does, a binary search may be performed between the parent block's timestamp and the wall clock time to find the time at which the first Subnet Validator gets removed.

Reviewer: ```suggestion
To ensure that validators are removed timely, blocks are considered valid if their timestamps are no greater than the time at which the first Subnet Validator gets removed from a lack of funds. This upholds the expectation that the number of Subnet Validators remains constant between blocks.

The block building protocol is modified to account for this change by first checking if the wall clock time removes any Subnet Validator due to a lack of funds. If the wall clock time does not remove any Subnet Validators, the wall clock time is used to build the block. If it does, the time at which the first Subnet Validator gets removed is used.
```

#### Fee Calculation

> Since $V-T$ is guaranteed to be constant between blocks, a simple formula can be derived. Prior to processing the next block, the total Subnet Validator fee assessed in the $\Delta t$ between the current block and the next block is::
> 
> $$\sum_{n=1}^{\Delta t} M \cdot \exp \left(\frac{x+n\cdot(V-T)}{K}\right)$$
> 
> However, this formula does not hold when $V-T < 0$ and $x < \Delta t \cdot (V-T)$. This is because the $x$ is nonnegative according to its update function $x = \max(x + \Delta t \cdot (V - T), 0)$. The above formula can be used during the time period where $x$ is decreasing by $T-V$ and is not floored to $0$:
> 
> $$\min\left(\Delta t, \left\lfloor{\frac{x}{T-V}}\right\rfloor\right)$$
> 
> For the remaining time period where $x$ would be floored to $0$, the fee is:
> 
> $$M \cdot \max\left(\Delta t - \left\lfloor{\frac{x}{T-V}}\right\rfloor, 0\right)$$
> 
> After the Subnet Validator fee is calculated, $x$ can be updated:
> 
> $$x = \max(x + \Delta t \cdot (V - T), 0)$$

Reviewer: ```suggestion
Prior to processing the next block, the total Subnet Validator fee assessed in the $\Delta t$ between the current block and the next block is:

```python
# Calculate the fee to charge over Δt
def cost_over_time(V:int, T:int, x:int, Δt: int) -> int:
    cost = 0
    for _ in range(Δt):
        x = max(x + V - T, 0)
        cost += fake_exponential(M, x, K)
    return cost
```

Reviewer: I couldn't figure out how to have the suggested change work correctly... But there needs to be a trailing "```" in the above suggested diff

#### Parameters

The parameters at activation are:

| Parameter | Value |
| - | - |
| $T$ | `10000` Subnet Validators |

> | $C$ | `20000` Subnet Validators |

Reviewer 2: ```suggestion
| $V$ | `20000` Subnet Validators |
```
